doc/write-cpp-unit-test.md

```diff
@@ -0,0 +1,276 @@
+# Wasm Extension C++ Unit Test Dev Guide
+
+> **NOTE:** This guide requires familiarity with [GoogleTest framework](https://github.com/google/googletest). Please go through its [getting started guide](https://github.com/google/googletest#getting-started) first.
+
+## Overview
+---
+
+In addition to [integration test](./write-integration-test.md), it is also recommended to write unit test for your extension. Unlike integration test, which executes the extension code as Wasm binary with Envoy runtime, unit test is compiled with local C++ toolchains and run natively.
+
+In [`proxy-wasm-cpp-host`](https://github.com/proxy-wasm/proxy-wasm-cpp-host), along with `V8`, `wasmtime`, `WAVM` runtimes, which could interpret Wasm binaries, [a `nullVM` runtime](https://github.com/proxy-wasm/proxy-wasm-cpp-host/tree/master/src/null) is also included, which could compile together with the extension code and run natively. The unit test will utilize `nullVm` runtime. Specifically, all host implementation of `proxy-wasm-cpp-host` (such as Envoy) needs to implements a [`Context` interface class](https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/eceb02d5b7772ec1cd78a4d35356e57d2e6d59bb/include/proxy-wasm/context.h#L128), which has many unimplemented host specific methods. In the unit test, a mock implementation of the `Context` class will be created using [GoogleTest framework](https://github.com/google/googletest), and compiled with your extension code. Within the mock implementation, we could inject desired host behavior and check if the extension interacts with host as expected.
```

Comment:
... runtimes that execute Wasm binaries... ... which runs natively compiled-in modules.